core/src/main/java/com/linecorp/armeria/common/logging/PrometheusMetricRequestDecorator.java

```diff
@@ -0,0 +1,351 @@
+                    }
+                    MetricLabel<?> label = labelingFunction.apply(log);
+                    metricAdapter.initialize(label, collectorRegistry)
+                                 .time(label, log.totalDurationNanos())
+                                 .responseBytes(label, log)
+                                 .decActiveRequests(label);
+
+                    if (isSuccess(log)) {
+                        metricAdapter.success(label);
+                    } else {
+                        metricAdapter.failure(label);
+                    }
+                }, RequestLogAvailability.COMPLETE);
+        return function.apply(ctx, req);
+    }
+
+    @Override
+    public O serve(Service<I, O> delegate, ServiceRequestContext ctx, I req) throws Exception {
+        return request(delegate::serve, ctx, req);
+    }
+
+    @Override
+    public O execute(Client<I, O> delegate, ClientRequestContext ctx, I req) throws Exception {
+        return request(delegate::execute, ctx, req);
+    }
+
+    /**
+     * Returns a new {@link Service} decorator that tracks request stats using the Prometheus metrics  library.
+     *
+     * @param collectorRegistry wrapped Prometheus registry
+     * @param labelingFunction function that will produce labels and values
+     * @param <I> Request type
+     * @param <O> Response type
+     * @return A service decorator function
+     */
+    public static <I extends Request, O extends Response> DecoratingServiceFunction<I, O> decorateService(
+            PrometheusRegistryWrapper collectorRegistry,
+            Function<RequestLog, MetricLabel> labelingFunction) {
+        return new PrometheusMetricRequestDecorator<>(collectorRegistry, labelingFunction);
+    }
+
+    /**
+     * Returns a new {@link Client} decorator that tracks request stats using the Prometheus metrics  library.
+     *
+     * @param collectorRegistry wrapped Prometheus registry
+     * @param labelingFunction function that will produce labels and values
+     * @param <I> Request type
+     * @param <O> Response type
+     * @return A client decorator function
+     */
+    public static <I extends Request, O extends Response> DecoratingClientFunction<I, O> decorateClient(
+            PrometheusRegistryWrapper collectorRegistry,
+            Function<RequestLog, MetricLabel> labelingFunction) {
+        return new PrometheusMetricRequestDecorator<>(collectorRegistry, labelingFunction);
+    }
+
+    private static boolean isSuccess(RequestLog log) {
+        if (log.responseCause() != null) {
+            return false;
+        }
+
+        if (HttpSessionProtocols.isHttp(log.sessionProtocol())) {
+            if (log.statusCode() >= 400) {
+                return false;
+            }
+        } else {
+            if (log.statusCode() != 0) {
+                return false;
+            }
+        }
+
+        final Object responseContent = log.responseContent();
+        if (responseContent instanceof RpcResponse) {
```

Comment:
As RpcResponse is created by business logic, it is the most precise, so should probably be the first check in this function.